### ISSUE TITLE
CrashLoopBackOff on mlab-staging maybe due to library change

### ISSUE THREAD

**Issue by stephen-soltesz** [m-lab/scraper#128]
### tl;dr

 * Apparently, two bad nodes were triggering CrashLoopBackOff errors in a handful of pods.
 * Draining those nodes, and deleting the pods still scheduled there ultimately addressed the issue.
 * As an aside, k8s recreates nodes that are deleted via GCP. But, that didn't help us here. Unclear why.
 * The root cause of the scraper exception is still unknown.

### Investigation

After finding that the pods on mlab-staging that were causing alerts appeared to be due to bad nodes, and after failing to recover them using the usual method of stopping or resetting the node, I checked the logs of the bad pods.

One did not have the stack trace, only the yacc warning. So those may be separate matters.

They seem to have a variation on the following:
```
$ kubectl -n scraper logs -c scraper fra02-mlab4-ndt-iupui-ndt-1149599371-rcr9s
WARNING: yacc table file version is out of date
Traceback (most recent call last):
  File "/run_scraper.py", line 36, in <module>
    import scraper
  File "/scraper.py", line 42, in <module>
    import google.cloud.datastore as cloud_datastore
  File "/usr/local/lib/python2.7/dist-packages/google/cloud/datastore/__init__.py", line 61, in <module>
    from google.cloud.datastore.client import Client
  File "/usr/local/lib/python2.7/dist-packages/google/cloud/datastore/client.py", line 28, in <module>
    from google.cloud.datastore._http import HTTPDatastoreAPI
  File "/usr/local/lib/python2.7/dist-packages/google/cloud/datastore/_http.py", line 19, in <module>
    from google.cloud import _http as connection_module
  File "/usr/local/lib/python2.7/dist-packages/google/cloud/_http.py", line 23, in <module>
    from google.cloud import exceptions
  File "/usr/local/lib/python2.7/dist-packages/google/cloud/exceptions.py", line 24, in <module>
    from google.api.core import exceptions
  File "/usr/local/lib/python2.7/dist-packages/google/api/core/exceptions.py", line 43, in <module>
    @six.python_2_unicode_compatible
AttributeError: 'module' object has no attribute 'python_2_unicode_compatible'
```

```
$ kubectl -n scraper get  pods | grep -v Running
NAME                                                       READY     STATUS             RESTARTS   AGE
den04-mlab4-neubot-mlab-neubot-3048705149-9t87x            1/2       CrashLoopBackOff   262        21h
dfw01-mlab1-ooni-mlab-ooni-704151753-wd2cr                 1/2       CrashLoopBackOff   7          21h
dfw02-mlab4-neubot-mlab-neubot-3228994741-bm5r3            1/2       CrashLoopBackOff   262        21h
dfw02-mlab4-npad-iupui-sidestream-262051259-b4nnj          1/2       CrashLoopBackOff   7          21h
dfw02-mlab4-utility-mlab-switch-1172738185-nd4zs           1/2       CrashLoopBackOff   7          21h
dfw03-mlab4-npad-iupui-npad-2953282528-4fqzn               1/2       CrashLoopBackOff   7          21h
dfw03-mlab4-npad-iupui-paris-traceroute-1945740849-bdk6w   1/2       CrashLoopBackOff   7          21h
dfw04-mlab4-ndt-iupui-ndt-1275756241-g6t8g                 1/2       CrashLoopBackOff   7          21h
dfw04-mlab4-neubot-mlab-neubot-3271658691-ssd4j            1/2       CrashLoopBackOff   7          21h
dfw04-mlab4-ooni-mlab-ooni-786923763-7rhv3                 1/2       CrashLoopBackOff   7          21h
dfw05-mlab4-npad-iupui-npad-2985460718-z3md7               1/2       CrashLoopBackOff   7          21h
dfw05-mlab4-ooni-mlab-ooni-2948301050-7dzks                1/2       CrashLoopBackOff   7          21h
fra01-mlab4-npad-iupui-paris-traceroute-1572513259-2j7m8   1/2       CrashLoopBackOff   7          21h
fra01-mlab4-npad-iupui-sidestream-2168690044-0mrzr         1/2       CrashLoopBackOff   7          21h
fra02-mlab4-ndt-iupui-ndt-1149599371-rcr9s                 1/2       CrashLoopBackOff   7          21h
fra03-mlab4-npad-iupui-paris-traceroute-1647158777-d9pkc   1/2       CrashLoopBackOff   7          21h
fra03-mlab4-utility-mlab-utilization-3529023358-cm048      1/2       CrashLoopBackOff   7          21h
fra04-mlab4-npad-iupui-npad-693666735-wpqzx                1/2       CrashLoopBackOff   261        21h
iad02-mlab4-ndt-iupui-ndt-3142680126-52drt                 1/2       CrashLoopBackOff   7          21h
lga03-mlab4-utility-mlab-switch-892178485-6xbck            1/2       CrashLoopBackOff   262        21h
lhr05-mlab4-utility-mlab-switch-1369870529-0k6xz           1/2       CrashLoopBackOff   262        21h
mia02-mlab4-utility-mlab-utilization-1286840169-lgndq      1/2       CrashLoopBackOff   262        21h
mia05-mlab4-npad-iupui-paris-traceroute-1647617529-f133g   1/2       CrashLoopBackOff   262        21h
nuq06-mlab4-ooni-mlab-ooni-3269493126-mdvdb                1/2       CrashLoopBackOff   262        21h
ord05-mlab4-neubot-mlab-neubot-1250594022-99n9n            1/2       CrashLoopBackOff   261        21h
par03-mlab4-ooni-mlab-ooni-2958852346-6wz1j                1/2       CrashLoopBackOff   262        21h
sea02-mlab4-utility-mlab-utilization-1356111735-h0d4v      1/2       CrashLoopBackOff   262        21h
sjc01-mlab4-utility-mlab-switch-1141346427-qfd4b           1/2       CrashLoopBackOff   262        21h
```

I deleted all pods.

Some stragglers were still having trouble on the same nodes.

I stopped nodes:
```
Node:           gke-scraper-cluster-scraper-node-pool-cee05ed3-3r72
Node:           gke-scraper-cluster-scraper-node-pool-cee05ed3-dgzk
```

Still a handful of issues, same stack trace in the logs. Pods are assigned to node '*-dgzk' above which should be 'Stopped'. I guess kubernetes restarts it.

```
$ kubectl -n scraper get  pods | grep -v Running
NAME                                                       READY     STATUS             RESTARTS   AGE
atl04-mlab4-ndt-iupui-ndt-3750527045-wchg8                 1/2       CrashLoopBackOff   3          5d
nuq03-mlab4-ndt-iupui-ndt-4007428291-hl7v5                 1/2       CrashLoopBackOff   3          5d
ord01-mlab4-neubot-mlab-neubot-4123954750-14d01            1/2       CrashLoopBackOff   3          5d
```

I've tried deleting both nodes via GCP as an experiment (one at a time). I didn't know how k8s would respond to that. Apparently k8s recreates it with the same GCE instance id.

But, that still didn't help entirely.

I drained the two problematic nodes, e.g.
```
kubectl drain --ignore-daemonsets gke-scraper-cluster-scraper-node-pool-cee05ed3-3r72
```

And then deleted pods. Seems to be okay now. All pods are running without issue.

Root cause of stacktrace is still unknown.

**Comment by stephen-soltesz** [comment]
Aaah, there was a travis initiated CRON build / deploy yesterday - https://travis-ci.org/m-lab/scraper/builds/269685632

**Comment by stephen-soltesz** [comment]
![cron](https://user-images.githubusercontent.com/1085316/29883056-068c1c8a-8d7d-11e7-84f6-2b70ba8c5360.png)

**Comment by pboothe** [comment]
Fixed in https://github.com/m-lab/scraper/pull/146

**Comment by pboothe** [comment]
The issue is now fixed.

So, here's our best guess at what happened:  cron caused an upload of a bad docker image that overwrote the good docker image. The tags were the same because the tags are based on the github commit hash and not the docker hash. The bad images crashlooped on the one node, but then the node getting cordoned off and drained caused them to get rescheduled on a node that had cached the old docker image and tags, and so did not download the new bad image with the same tags, and so began to work.

The root cause is CRON + the underlying image getting updated in a way that broke our build + not testing the code INSIDE the image.  We now test the code inside the image, which means that breaking library or environment changes will now result in failed tests, which should prevent this problem from re-occurring.

https://github.com/m-lab/scraper/pull/146 fixed the issue and fixed the unit tests to detect the issue in the future.
